### ISSUE TITLE
team_build_events_NNN and pipeline_build_events_NNN tables are not cleaned up

### ISSUE THREAD

**Issue by vito** [concourse/concourse#2169]
This is demonstrated by our test deployment in the pipeline. Running `fly teams` shows only a few teams, while hundreds are still in the database.

**Comment by timrchavez** [comment]
I feel like we're seeing this, but I also thought it was maybe addressed. Any updates?

**Comment by vito** [comment]
There was a recent regression causing build events to not be cleaned up, but that's since been fixed and shipped. This issue is meant for the tables themselves, which don't seem to be deleted when the team or pipeline is destroyed.

**Comment by christophermancini** [comment]
Do you think it is safe to just drop tables that have zero rows and are likely orphaned?

**Comment by christophermancini** [comment]
How soon will you have a fix? Our installation is struggling hardcore.

**Comment by vito** [comment]
It's struggling due to empty tables?

**Comment by nelsam** [comment]
@vito I'm too willing to throw caution to the wind, apparently :)  We had over 16000 child tables for `build_events`, and since none of those children had constraints on them, any queries against `build_events` (namely: the gc cleanup that runs `delete from build_events where build_id in (so, very, many, variables) would result in over 16000 index scans (and a ton of variable binding).

We wrote a pl/pgsql function that unlinked any tables that appeared to be orphans and everything's back to normal.

**Comment by timrchavez** [comment]
:point_up:  Which would effectively eat 64GB of RAM in a few seconds. The only thing preventing us from OOM'ing the postgres process, itself, was using an overcommit limit in the kernel (but that had the affect of OOM'ing queries)

**Comment by vito** [comment]
Interesting! Alright, if it helps put the fires out go ahead. I can't
promise that it'll be a smooth transition once the fix is out but I'm happy
to help fix any problems it causes manually.

**Comment by nelsam** [comment]
fwiw, here's the function we ran:
```sql
create or replace function unlink_orphaned_build_events_tables() RETURNS void AS $$
DECLARE
  t RECORD;
BEGIN
  FOR t IN SELECT relname FROM pg_stat_user_tables tbls LEFT OUTER JOIN teams ON tbls.relname = 'team_build_events_' || teams.id::varchar LEFT OUTER JOIN pipelines ON tbls.relname = 'pipeline_build_events_' || pipelines.id::varchar WHERE (relname LIKE 'team_build_events_%' AND teams.id IS NULL) OR (relname LIKE 'pipeline_build_events_%' AND pipelines.id IS NULL) LOOP
    RAISE NOTICE 'unlinking table %', t.relname;
    EXECUTE 'ALTER TABLE ' || quote_ident(t.relname) || ' NO INHERIT build_events';
  END LOOP;
END;
$$ LANGUAGE plpgsql;
CREATE FUNCTION
```

**Comment by timrchavez** [comment]
@alpha-er Load was slightly higher. The main concern was memory pressure.

**Comment by alpha-er** [comment]
@timrchavez we're currently having problems with our concourse database where the cpu usage is constantly increasing. I thought that this issue could be related to that but as you said the main concern is on memory. We recently upgraded our RDS instance. Here's a graph of the db's cpu usage:
<img width="946" alt="screen shot 2018-07-04 at 4 50 07 pm" src="https://user-images.githubusercontent.com/4862363/42257374-69b56102-7faa-11e8-9b1f-44dd62a138a9.png">

- The drop at the start was when we upgraded the RDS instance.
- The spike in the middle was when there were two web nodes running.
- The slight decrease at the end was when performing `vacuum analyze` on concourse tables.

**Comment by pivotal-jwinters** [comment]
We noticed a couple things to fix.. and fixed them!
- the `team_build_events_*` tables get truncated but not deleted when a team is deleted; we now drop the table instead of truncating.
- when a team which owns a specific pipeline is deleted, the pipeline entry is cascade deleted, but the `pipeline_build_events_*` table lives on; we now get the list of pipelines owned by the team and delete their `pipeline_build_events_*` along with the team.

**Comment by alpha-er** [comment]
@pivotal-jwinters what would that mean for `team_build_events_*` and `pipeline_build_events_*` of teams that were already deleted prior to the version upgrade where this fix will be included?

**Comment by pivotal-jwinters** [comment]
@alpha-er we're actually trying a different approach using `triggers`, instead of manually adding/deleting the corresponding tables.

Either way, we'll probably end up writing a migration to purge all the orphaned tables.

**Comment by vito** [comment]
We decided not to reap existing tables in the migration, as it feels too dangerous and complicated to perform automatically.

I would recommend manually dropping the orphaned tables at a time when no new teams or pipelines are going to be created/destroyed.

I can confirm though that there's no build-up of tables on our periodic test deployments, so I'm gonna close this out! Along the way I noticed a bunch of dangling postgres sequence objects, which we could maybe clean up with triggers as part of a later story as it's just about the exact same situation.
